### PR TITLE
HeavyFlavor Validation update

### PR DESCRIPTION
**Pull request by tmadlener**
Necessary updates to have meaningful/filled validation plots again for the HeavyFlavor validation package (used by BPH for RelVal):

- Update the trigger paths to the 2017 HLT menu. Where possible the paths from 2016 have been updated to their 2017 version. Some paths don't have a 2017 version. These have been removed.
- Refactored the assignment of filter levels to the different filters (based on the filter name) to make it more stable and to get rid of newly introduced duplicates due to the IterL3 modules in the paths.

### CODE REVIEW

**Comment by mtosi** [comment]
@cmsbuil, please test

**Comment by mtosi** [comment]
@tmadlener first of all, thanks !
then, may I suggest to update the directory into BPH ?

**Comment by tmadlener** [comment]
@mtosi which directory are you referring to? The output directory of the plots or the package directory inside cmssw?

The first one is no problem. Only one parameter in the config file has to be changed for that and according to my quick check there is currently no conflicting BPH folder in the output directory.

The second one is a bit more work, but should be possible as well. However, I would have to check how many files would have to be changed and some of them probably would also need to be renamed. You can probably judge better than me how much additional work would be involved in that (just the renaming, etc. part).

**Comment by mtosi** [comment]
sorry the confusion
I mean

baseFolderPath = 'HLT/HeavyFlavor/HLT/'
 -->

baseFolderPath = 'HLT/BPH/'

thanks
 mia

**Comment by dmitrijus** [comment]
+1

**Comment by tmadlener** [comment]
Hi all,
I have a question about the backports. Will this PR be automatically backported to 9_2_X or is this something we have to request (or do ourselves)?

Cheers, Thomas

**Comment by mtosi** [comment]
there are not "automatic backport"
  mia

**Comment by tmadlener** [comment]
I see. So to clarify. What are the actions that have to be taken from our side to have a backport of this PR?

**Comment by dmitrijus** [comment]
You have to make a pull request to the target branch.

And you should provide a good reason to have it backported - the PR will be touching production code.

**Comment by tmadlener** [comment]
Thanks for the prompt answers.
Another question: Will this PR be in the next 9_3_0 pre-release (if there is one) or is a backport to 9_3_X necessary as well for that?

**Comment by mtosi** [comment]
for the TSG is not mandatory
(but I would suggest to do so)

**Comment by tmadlener** [comment]
Ok, thanks for your patience and answers. :)
Since there are probably more validation campaigns to come for 9_2_X and 9_3_X releases, I will open PRs targetting these two branches as soon as this one has been merged, if there are no objections.

Having these changes also in those branches allows us to use the DQM GUI again for upcoming validation campaigns. Currently we have to run the validation and harvesting modules manually for each release, which gets tedious.

**Comment by dmitrijus** [comment]
+1

File: HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc [hunk 1]
```diff
@@ -145,43 +158,29 @@ void HeavyFlavorValidation::dqmBeginRun(const edm::Run& iRun, const edm::EventSe
...
 		}
 	}
-	if(filterNamesLevels.size()==0){
+
+	if(filterNamesLevels.empty()){
 		LogDebug("HLTriggerOfflineHeavyFlavor")<<"Bad Trigger Path: "<<triggerPathName<<endl;
 		return;
 	}else{
+          	stringstream os;
+                for (const auto &filters : filterNamesLevels) os << " " << filters.first;
```
Comment (davidlange6) [inline HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc]: hi @tmadlener - you can avoid the string stream here - just str=str+" "+filters.first in the loop

File: HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc [hunk 2]
```diff
@@ -649,6 +648,71 @@ void HeavyFlavorValidation::myBook1D(DQMStore::IBooker & ibooker, TString name,
...
+                              "HLTMuonTrackMassFilter"};
+  const std::vector<std::string> l5Filters = {"HLTmumutkFilter", "HLT2MuonMuonDZ", "HLTDisplacedmumuFilter"};
+
+  if (contains(moduleName, "Filter") && hltConfig.moduleEDMType(moduleName) == "EDFilter") {
+    int level = -1; // if we get here, one of the following clauses should always catch!
+    if (contains(moduleName, "L1") && !contains(moduleName, "ForIterL3") &&
+        hltConfig.moduleType(moduleName) == l1Filter) {
+      level = 1;
+    }
+    if (contains(moduleName, "L2") && hltConfig.moduleType(moduleName) == l2Filter) {
```
Comment (davidlange6) [inline HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc]: can this and subsequent if statements be else if instead? eg, once you have found a level value, the subsequent ifs do not need to be checked, right?

**Comment by davidlange6** [comment]
please test


On 25 Sep 2017, at 09:49, cmsbuild <notifications@github.com<mailto:notifications@github.com>> wrote:


Pull request #20520<https://github.com/cms-sw/cmssw/pull/20520> was updated. @kmaeshima<https://github.com/kmaeshima>, @cmsbuild<https://github.com/cmsbuild>, @vanbesien<https://github.com/vanbesien>, @vazzolini<https://github.com/vazzolini>, @dmitrijus<https://github.com/dmitrijus> can you please check and sign again.

**Comment by davidlange6** [comment]
merge

**Pull request merged**